build.gradle

```diff
@@ -235,76 +236,99 @@ java {
+            'src/gameTest/resources/META-INF/gametest_ats.cfg',
+    ]
 
-        for (String name : secondaryModules) {
-            modSources.add(getExtraSourceSet(name, 'datagen').get())
-            programArguments.addAll((String[]) ['--mod', "mekanism${name}", '--existing', file("src/${name}/resources/").absolutePath])
-        }
+    runs {
+        configureEach {
+            if (hasProperty('forge_force_ansi')) {
+                //Force ansi if declared as a gradle variable, as the auto detection doesn't detect IntelliJ properly
+                // or eclipse's plugin that adds support for ansi escape in console
+                systemProperties.put('terminal.ansi', (String) property('forge_force_ansi'))
+            }
+
+            loadedMods = [mods.mekanism]
+            for (String name : secondaryModules) {
+                loadedMods.add(mods.named(name))
+            }
 
-        dependencies {
-            runtime(configurations.datagenNonMod)
+            //if the selected toolchain is a JBR, enable DCEVM
+            if (javaToolchains.launcherFor(java.toolchain).map { it.metadata.vendor }.getOrElse('').contains('JetBrains')) {
+                jvmArguments.addAll('-XX:+IgnoreUnrecognizedVMOptions', '-XX:+AllowEnhancedClassRedefinition')
+            }
+        }
+        //Note: To enable logging into the client account, set the enable_devlogin property to true in your gradle user home
+        client {
+            client()
+            devLogin = Boolean.valueOf(project.enable_devlogin)
+        }
+        clientAlt {
+            client()
+            // Force dev login to be disabled for this run to allow using a secondary username
+            devLogin = false
+            programArguments.addAll('--username', 'AltDev')
+        }
+        server {
+            server()
+        }
+        gameTestServer {
+            type = "gameTestServer"
+            sourceSet = sourceSets.gameTest
+            loadedMods = neoForge.mods
+        }
+        gameTestClient {
+            client()
+            sourceSet = sourceSets.gameTest
+            loadedMods = neoForge.mods
+        }
+        data {
+            data()
+            programArguments.addAll('--all', '--output', file('src/datagen/generated/').absolutePath,
+                                    '--mod', 'mekanism', '--existing', file('src/main/resources/').absolutePath)
+
+            sourceSet = sourceSets.datagenMain
+            loadedMods.add(mods.mekanism)
+
+            for (String name : secondaryModules) {
+                println("Adding $name")
+                loadedMods.add(mods.named(name))
+                programArguments.addAll('--mod', 'mekanism' + name, '--existing', file("src/${name}/resources/").absolutePath)
+            }
         }
     }
+
+    unitTest {
+        enable()
+        testedMod = mods.mekanism
+    }
 }
 
+configurations.dataAdditionalRuntimeClasspath.extendsFrom configurations.datagenNonMod
```

Comment:
Why is this down here instead of the normal configurations block, and what is the purpose of this?